cranelift-codegen/meta/src/cdsl/instructions.rs

```diff
@@ -406,36 +418,102 @@ impl ValueTypeOrAny {
     }
 }
 
+type VectorBitWidth = u64;
+pub enum InstructionParameter {
+    AnyType,
+    LaneType(LaneType),
+    VectorType(LaneType, VectorBitWidth),
+    ReferenceType(ReferenceType),
+}
+
+impl From<shared::types::Int> for InstructionParameter {
+    fn from(ty: shared::types::Int) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
+impl From<shared::types::Bool> for InstructionParameter {
+    fn from(ty: shared::types::Bool) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
+impl From<shared::types::Float> for InstructionParameter {
+    fn from(ty: shared::types::Float) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
+impl From<cdsl::types::LaneType> for InstructionParameter {
+    fn from(ty: cdsl::types::LaneType) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
 #[derive(Clone)]
 pub struct BoundInstruction {
     pub inst: Instruction,
     pub value_types: Vec<ValueTypeOrAny>,
 }
 
 impl BoundInstruction {
-    pub fn bind(self, lane_type: impl Into<LaneType>) -> BoundInstruction {
-        bind(self.inst, Some(lane_type.into()), self.value_types)
-    }
-
-    pub fn bind_ref(self, reference_type: impl Into<ReferenceType>) -> BoundInstruction {
-        bind_ref(self.inst, Some(reference_type.into()), self.value_types)
+    /// Construct a new bound instruction (with nothing bound yet) from an instruction
+    fn new(inst: &Instruction) -> Self {
+        BoundInstruction {
+            inst: inst.clone(),
+            value_types: vec![],
+        }
     }
 
-    pub fn bind_vector_from_lane(
-        self,
-        lane_type: impl Into<LaneType>,
-        vector_size_in_bits: u64,
-    ) -> BoundInstruction {
-        bind_vector(
-            self.inst,
-            lane_type.into(),
-            vector_size_in_bits,
-            self.value_types,
-        )
+    /// Verify that binding types to the instruction does not violate the polymorphic rules.
+    fn verify_polymorphic_binding(&self) -> Result<(), String> {
```

Comment:
I can make that change but I would prefer to keep it as a `Result` because there's minimal performance concern during meta, it's easier to test, and I asked @sunfishcode at some point about how you guys want to do error-handling in cranelift and he expressed a preference to moving to `Result`s.

Reply:
It wasn't a performance concern from my point of view. I was proposing that would simplify reading and understanding this code in my opinion, but this is highly subjective. I don't have a strong opinion here.